c2rust-analyze/src/rewrite/expr/mir_op.rs

```diff
@@ -485,51 +502,170 @@ impl<'a, 'tcx> ExprRewriteVisitor<'a, 'tcx> {
-        {
-            // TODO: emit rewrite
-            from.qty = to.qty;
+        // Early `Ownership` casts.  We do certain casts here in hopes of reaching an `Ownership`
+        // on which we can safely adjust `Quantity`.
+        while from.own != to.own {
+            match self.cast_ownership_one_step(from, to, true) {
+                Some(new_own) => {
+                    from.own = new_own;
+                }
+                None => break,
+            }
         }
 
-        if matches!(
-            from.qty,
-            Quantity::Array | Quantity::OffsetPtr | Quantity::Slice
-        ) && to.qty == Quantity::Single
-        {
-            let mutbl = match from.own {
-                Ownership::Imm => Some(false),
+        // Safe casts that change `Quantity`.
+        while from.qty != to.qty {
+            // Mutability of `from`.  `None` here means that safe `Quantity` conversions aren't
+            // possible given `from`'s `Ownership`.  For example, we can't convert `Box<[T]>` to
+            // `Box<T>`.
+            let opt_mutbl = match from.own {
+                // Note that `Cell` + `Slice` is `&[Cell<T>]`, not `&Cell<[T]>`, so it can be
+                // handled like any other `&[_]`.
+                Ownership::Imm | Ownership::Cell => Some(false),
                 Ownership::Mut => Some(true),
-                // Note that `Cell` + `Slice` is `&[Cell<T>]`, not `&Cell<[T]>`.
-                Ownership::Cell => Some(false),
                 _ => None,
             };
-            if let Some(mutbl) = mutbl {
-                self.emit(RewriteKind::SliceFirst { mutbl });
-                from.qty = Quantity::Single;
-            }
-        }
+            match (from.qty, to.qty) {
+                (Quantity::Array, _) => {
+                    // `Array` goes only to `Slice` directly.  All other `Array` conversions go
+                    // through `Slice` first.
+                    error!("NYI: cast Array to {:?}", to.qty);
+                    from.qty = Quantity::Slice;
+                }
+                // Bidirectional conversions between `Slice` and `OffsetPtr`.
+                (Quantity::Slice, Quantity::OffsetPtr) => {
+                    let _rw = match opt_mutbl {
+                        // Currently a no-op, since `Slice` and `OffsetPtr` are identical.
+                        Some(_mutbl) => (),
+                        None => break,
+                    };
+                    // self.emit(rw);
+                    from.qty = Quantity::OffsetPtr;
+                }
+                (Quantity::OffsetPtr, Quantity::Slice) => {
+                    let _rw = match opt_mutbl {
+                        // Currently a no-op, since `Slice` and `OffsetPtr` are identical.
+                        Some(_mutbl) => (),
+                        None => break,
+                    };
+                    // self.emit(rw);
+                    from.qty = Quantity::Slice;
+                }
```

Comment:
Could these be combined since this is a bidirectional conversion?

```suggestion
                (Quantity::Slice, Quantity::OffsetPtr) | (Quantity::OffsetPtr, Quantity::Slice) => {
                    let _rw = match opt_mutbl {
                        // Currently a no-op, since `Slice` and `OffsetPtr` are identical.
                        Some(_mutbl) => (),
                        None => break,
                    };
                    // self.emit(rw);
                    from.qty = to.qty;
                }
```

Also, why is the `_rw` still needed?  Even if it's used, it's `()`.